### PR TITLE
changes to configure DTA to handle infinite reboot problem

### PR DESCRIPTION
**Pull request by vimegh**
Bug 922874: [DTA] Validate the necessary prereqs for DTA configuration

Description: Check for Auologon and legal notice enable and issue warnings. Handle infinite reboot scenarios

Changes done: 
1. Registry checks added to check for legalnotice and autologon and add necessary log messages for warning to be shown
2. Added Update-RebootCount function which save reboot count and dtlenvironment url in registry. Configure test agent throws a error if this count has reached a value more than 3. This count is increased before returning restart code.

Tested for infinite reboot scenario. Failing after 3 restarts.  Also adding necessary log messages.

### CODE REVIEW

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 [hunk 1]
```diff
@@ -1,4 +1,135 @@
...
+	}
+    elseif($autoadminLogon -eq "1")
+    {
+	    return $false
+    }
+
+    return $true
+}
+
+function isShowLogonMessagePolicyEnabled()
```
Comment (cltshivash) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: Please fix the formatting.. add logs where needed..

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 [hunk 2]
```diff
@@ -801,7 +955,7 @@ function ConfigurePowerOptions([System.Management.Automation.PSCredential] $Mach
 function CreateNewSession([System.Management.Automation.PSCredential] $MachineCredentials)
 {
     Write-Verbose -Message("Trying to fetch WinRM details on the machine") -Verbose
-    $wsmanoutput = Get-WSManInstance –ResourceURI winrm/config/listener –Enumerate
+    $wsmanoutput = Get-WSManInstance -ResourceURI winrm/config/listener -Enumerate
```
Comment (cltshivash) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: please check this..

**Comment by acesiddhu** [comment]
bump up task version

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 [hunk 3]
```diff
@@ -1,4 +1,135 @@
...
+        $registryPath = "HKLM:\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Winlogon"
+    }
+
+    if (-not (Test-Path $registryPath))
+	{
+		return $true
+	}
+	
+	$autoadminLogon = (Get-ItemProperty $registryPath -ErrorAction SilentlyContinue).AutoAdminLogon
+	if (($autoadminLogon -eq $null) -or ($autoadminLogon.Length -eq 0))
```
Comment (nigurr) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: if($autoadminLogon)

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 [hunk 4]
```diff
@@ -919,22 +1075,27 @@ function ConfigureTestAgent
...
         Write-Verbose -Message ("TestAgent Configured Successfully") -Verbose
     }
     elseif($retCode -eq 3010)
     {
         Write-Verbose -Message ("TestAgent configuration requested for reboot") -Verbose
+        $rebootCount = Update-RebootCount($EnvironmentUrl)
+        if($rebootCount -gt 3)
+        {
+            $retCode = 1
+            throw ("Exceeded maximum number of reboots. TestAgent Configuration is failing with exit code {0}. Error code : {1}" -f $LASTEXITCODE, $retCode)
```
Comment (nigurr) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: >  throw ("Exceeded maxim [](start = 10, length = 24)

throwing error may not use retCode. check

**Comment by nigurr** [comment]
:shipit:

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 [hunk 5]
```diff
@@ -1,4 +1,135 @@
-﻿function Get-TestAgentType([string] $Version)
+﻿function isAutoLogonDisabled()
+{
+    #check 64 bit or 32 bit
+	if ([IntPtr]::Size -eq 8)
+	{
```
Comment (nigurr) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: go for Test-Path for both

Comment (vimegh) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: done

---
In reply to: [103681498](https://github.com/Microsoft/vsts-tasks/pull/3701#discussion_r103681498) [](ancestors = 103681498)

**Comment by vimegh** [comment]
Task version will be updated by Kavipriya. We don't want multiple task version updates as it might create problem for user in future when they update from 2015U4 to 2017. 

---
In reply to: [283323311](https://github.com/Microsoft/vsts-tasks/pull/3701#issuecomment-283323311) [](ancestors = 283323311)

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 [hunk 6]
```diff
@@ -1,4 +1,143 @@
...
+	else
+	{
+		$testAgentRegPath = "HKLM:\SOFTWARE\Microsoft\TestAgentConfig" 
+	}
+
+	if (-not (Test-Path $testAgentRegPath))
+	{
+		New-Item -Path $testAgentRegPath -Force | Out-Null
+		Write-Verbose -Message ("Updating machine reboot count to 1") -Verbose
+		New-ItemProperty -Path $testAgentRegPath -Name "MachineRebootCount" -Value 1 -PropertyType DWord -Force | Out-Null
```
Comment (acesiddhu) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: when are we deleting these entries?

Comment (vimegh) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: we are resetting reboot count every time there is new environment URL, and same key and name is being reused so no need to delete

---
In reply to: [103856036](https://github.com/Microsoft/vsts-tasks/pull/3701#discussion_r103856036) [](ancestors = 103856036)

**Pull request merged**